pybricks/tools/pb_type_hostbuffer.c

```diff
@@ -0,0 +1,118 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2024 The Pybricks Authors
+
+#include "py/mpconfig.h"
+
+#if PYBRICKS_PY_TOOLS_HOSTBUFFER
+
+#include <string.h>
+
+#include <pbsys/command.h>
+
+#include "py/mphal.h"
+#include "py/objstr.h"
+
+#include <pybricks/tools.h>
+
+#include <pybricks/util_mp/pb_kwarg_helper.h>
+#include <pybricks/util_mp/pb_obj_helper.h>
+#include <pybricks/util_pb/pb_error.h>
+
+typedef struct _pb_type_hostbuffer_obj_t {
+    mp_obj_base_t base;
+    mp_obj_t format;
+    mp_obj_str_t bytes_obj;
+    uint8_t buffer[];
+} pb_type_hostbuffer_obj_t;
+
+// pointer to dynamically allocated hostbuffer singleton for driver callback.
+static pb_type_hostbuffer_obj_t *hostbuffer_instance;
+
+static void handle_write_data_buffer(uint16_t offset, uint32_t size, const uint8_t *data) {
+    // Can't write if buffer does not exist or isn't big enough.
+    if (!hostbuffer_instance || offset + size > hostbuffer_instance->bytes_obj.len) {
+        return;
+    }
+    memcpy(hostbuffer_instance->buffer + offset, data, size);
+}
+
+STATIC mp_obj_t pb_type_hostbuffer_get_bytes(mp_obj_t self_in) {
```

Comment:
To keep things a simple as possible, maybe we don't need this method?

Reply:
I suppose that works. Giving full access to the bytes seemed more generic, i.e. the host could even dictate the format if it is unknown and send it as a zero terminated string at the start of the buffer.